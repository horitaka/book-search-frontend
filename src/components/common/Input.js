import React, {useState} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import * as Colors from '../../constants/Colors';
import * as Fonts from '../../constants/Fonts';

function Input(props) {
  const {type, placeholder, onChange, flex} = props;
  const [value, setValue] = useState('')

  function handleKeywordChange(event) {
    const value = event.target.value;
    setValue(value);
    onChange(value)
  }

  return (
    <StyledInput
      type={type || "text"}
      placeholder={placeholder}
      value={value}
      flex={flex}
      onChange={handleKeywordChange}
    />
  )
}

Input.propTypes = {
  type: PropTypes.string,
  placeholder: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  flex: PropTypes.string,
}

const StyledInput = styled.input`

  flex: ${props => props.flex || '1 0 auto'};
  margin: 0;
  padding: 5px;
  border: 1px solid ${Colors.primaryColor};
  border-radius: ;
  :focus {
    border: 1px solid ${Colors.accentColor};
    outline: 0;
  }

  font: ${Fonts.fontFamily};
  font-size: ${Fonts.fontLarge};
`

export default Input;
