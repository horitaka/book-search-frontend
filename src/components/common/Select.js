import React, { useState} from 'react'
import PropTypes from 'prop-types';
import styled from 'styled-components';

import * as Colors from '../../constants/Colors';
import * as Fonts from '../../constants/Fonts';

function Select(props) {
  const {options, onSelected, flex, margin} = props;

  const defaultOption = options.find(option => option.isDefault);
  const defaultValue = defaultOption ? defaultOption.value : '';
  const [selectedOption, setOption] = useState(defaultValue);

  function handleSelect(event) {
    const item = event.target.value;
    setOption(item)
    onSelected(event)
  }

  return(
    <StyledSelect value={selectedOption} onChange={handleSelect} flex={flex} margin={margin}>
      { options.map( option => <option  disabled={option.disabled} value={option.value} key={option.value}>{option.label}</option>)}
    </StyledSelect>
  )
}

Select.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      disabled: PropTypes.bool,
      isDefault: PropTypes.bool,
    })).isRequired,
  onSelected: PropTypes.func.isRequired,
  flex: PropTypes.string,
}

const StyledSelect = styled.select`
  box-sizing: border-box;

  flex: ${props => props.flex || '1 0 auto'};
  width: auto;
  margin: ${props => props.margin || 0};
  padding: 5px;
  background-color: transparent;
  border-radius: 3px;
  outline: none;
  border: slod 1px ${Colors.primaryColor};

  :active {
    background: ${Colors.secondoryColor};
    outline: none;
  }

  font-size: ${Fonts.fontMedium}px;
  color: black;
`


export default Select;
