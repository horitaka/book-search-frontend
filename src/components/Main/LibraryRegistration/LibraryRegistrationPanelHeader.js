import React, {useEffect} from 'react'
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';

import Select from '../../common/Select';
import Input from '../../common/Input';
import Container from '../../common/Container';
import Box from '../../common/Box';
import { prefectureList } from './PrefectureList'

function LibraryRegistrationPanelHeader({onPrefSelected, onKeywordChanged}) {
  const defaultPrefecutre = '東京都'
  let prefectureOptions = prefectureList.map(prefecture => {
    const isDefault = prefecture === defaultPrefecutre;
    return ({
      value: prefecture,
      label: prefecture,
      isDefault: isDefault,
    })
  })

  const onKeywordChangedDebounced = debounce(onKeywordChanged, 500);

  useEffect(() => {
    onPrefSelected(defaultPrefecutre)
  }, [onPrefSelected]);

  function handlePrefSelected(event) {
    const prefecture = event.target.value;
    onPrefSelected(prefecture)
  }

  function handleKeywordChange(keyword) {
    onKeywordChangedDebounced(keyword);
  }

  return(
    <Box flex={'0 1 auto'}>
      <Container direction={'row'} justifyContent={'center'} alignItems={'center'}>
        <Select options={prefectureOptions} onSelected={handlePrefSelected} margin={'0 5px 0 0'}/>
        <Input type="text" placeholder="図書館を検索..." onChange={handleKeywordChange}/>
      </Container>
    </Box>
  )
}

LibraryRegistrationPanelHeader.propTypes = {
  onPrefSelected: PropTypes.func.isRequired,
  onKeywordChanged: PropTypes.func.isRequired,
}


export default LibraryRegistrationPanelHeader;
