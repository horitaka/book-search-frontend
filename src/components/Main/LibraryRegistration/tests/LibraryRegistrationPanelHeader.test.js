import React from 'react';
import { shallow } from 'enzyme';

import LibraryRegistrationPanelHeader from '../LibraryRegistrationPanelHeader'
import Select from '../../../common/Select'
import Input from '../../../common/Input';

jest.mock('lodash.debounce', () => jest.fn(fn => fn))

describe('<LibraryRegistrationPanelHeader> Component', () => {
  const onPrefSelectedMock = jest.fn();
  const onKeywordChangedMock = jest.fn();
  const events = {
    onPrefSelected: onPrefSelectedMock,
    onKeywordChanged: onKeywordChangedMock
  }

  it('render correctly', () => {
    const wrapper = shallow(<LibraryRegistrationPanelHeader {...events}/>)
    expect(wrapper).toMatchSnapshot();
  });

  describe('when props is normal', ()=> {
    it('render Select with option props', () => {
      const prefectureOptionDefault = {
        value: '東京都',
        label: '東京都',
        isDefault: true,
      };

      const prefectureOptionNotDefault = {
        value: '神奈川県',
        label: '神奈川県',
        isDefault: false,
      }

      const wrapper = shallow(<LibraryRegistrationPanelHeader {...events}/>)

      expect(wrapper.find(Select).prop('options')).toContainEqual(prefectureOptionDefault);
      expect(wrapper.find(Select).prop('options')).toContainEqual(prefectureOptionNotDefault);
    });

    it('', () => {

    });
  });

  describe('when onSelect is called', ()=> {
    it('call onPrefSelected', () => {
      const wrapper = shallow(<LibraryRegistrationPanelHeader {...events}/>)
      const event = {
        target: {
          value: '東京都',
        }
      }

      expect(onPrefSelectedMock).toHaveBeenCalledTimes(0);
      wrapper.find(Select).simulate('selected', event);
      expect(onPrefSelectedMock).toHaveBeenCalledTimes(1);
      expect(onPrefSelectedMock).toHaveBeenCalledWith('東京都');
    });

  });

  describe('when Input value is changed', ()=> {
    it('call onKeywordChanged', () => {
      const wrapper = shallow(<LibraryRegistrationPanelHeader {...events}/>)
      const keyword = '東京都'

      expect(onKeywordChangedMock).toHaveBeenCalledTimes(0);
      wrapper.find(Input).simulate('change', keyword);
      expect(onKeywordChangedMock).toHaveBeenCalledTimes(1);
      expect(onKeywordChangedMock).toHaveBeenCalledWith('東京都');
    });

  });
})
