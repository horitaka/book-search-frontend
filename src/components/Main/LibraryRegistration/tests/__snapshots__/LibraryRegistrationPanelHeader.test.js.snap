// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LibraryRegistrationPanelHeader> Component render correctly 1`] = `ShallowWrapper {}`;
